community/detectors/gitlab_cve_2021_22205/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202122205/Cve202122205VulnDetector.java

```diff
@@ -0,0 +1,230 @@
+    StringBuilder targetUrlBuilder = new StringBuilder();
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      targetUrlBuilder
+          .append("http://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    return targetUrlBuilder;
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("CVE-2921-22205 starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(Cve202122205VulnDetector::isWebServiceOrUnknownService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private String clearCookie(String cookie) {
+    return cookie.split("path=/;")[0];
+  }
+
+  private Cve202122205VulnVo getCsrfTokenAndCookie(NetworkService networkService) {
+    String targetUserSignUrl = buildTarget(networkService).append(USER_SIGN_PATH).toString();
+    Cve202122205VulnVo result = new Cve202122205VulnVo();
+    try {
+      HttpResponse httpResponse =
+          httpClient.send(get(targetUserSignUrl).withEmptyHeaders().build(), networkService);
+      if (httpResponse.status().code() == 200) {
+        String cookie = httpResponse.headers().get(SET_COOKIE).get();
+        Matcher csrfTokenMatcher = CSRF_TOKEN_PATTERN.matcher(httpResponse.bodyString().get());
+        if (csrfTokenMatcher.find() && !cookie.isEmpty()) {
+          result.setCsrfToken(csrfTokenMatcher.group(1));
+          StringBuilder cookies = new StringBuilder();
+          Iterator<String> headerCookies =
+              httpResponse.headers().getAll(SET_COOKIE).stream().iterator();
+          while (headerCookies.hasNext()) {
+            cookies.append(clearCookie(headerCookies.next()));
+          }
+          result.setCookie(cookies.toString());
+        }
+      }
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+      return null;
+    }
+    return result;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetVulnerabilityUrl = buildTarget(networkService).append(VUL_PATH).toString();
+    try {
+      Cve202122205VulnVo info = getCsrfTokenAndCookie(networkService);
+      assert info != null;
+      if (Strings.isNullOrEmpty(info.getCookie()) && Strings.isNullOrEmpty(info.getCsrfToken())) {
+        logger.atWarning().log("Get %s Csrf Token and Cookie failed", networkService);
+        return false;
+      }
+      byte[] payload = BaseEncoding.base16().decode(POST_DATA.toUpperCase(Locale.ROOT));
```

Comment:
`.toUpperCase` is not needed since you've already converted the content of `POST_DATA` to all upper case.